pkgs/applications/networking/geph/default.nix

```diff
@@ -50,52 +54,101 @@ in
     src = fetchFromGitHub {
       owner = "geph-official";
       repo = "gephgui-pkg";
-      rev = "85a55bfc2f4314d9c49608f252080696b1f8e2a9";
-      hash = "sha256-id/sfaQsF480kUXg//O5rBIciuuhDuXY19FQe1E3OQs=";
+      rev = "1974ce95b908cb16dafa7c19f732ae27ff2928a8";
+      hash = "sha256-4BVplOEBm4kGhG8Dc4uBv76s2Y3wDxHsZQdpvro9hUY=";
       fetchSubmodules = true;
     };
 
-    gephgui = buildNpmPackage {
-      pname = "gephgui";
-      inherit version src;
-
-      sourceRoot = "${src.name}/gephgui-wry/gephgui";
+    pnpm-deps = stdenvNoCC.mkDerivation {
+      pname = "${pname}-pnpm-deps";
+      inherit src version;
 
-      postPatch = "ln -s ${./package-lock.json} ./package-lock.json";
+      sourceRoot = "source/gephgui-wry/gephgui";
 
-      npmDepsHash = "sha256-5y6zpMF4M56DiWVhMvjJGsYpVdlJSoWoWyPgLc7hJoo=";
+      nativeBuildInputs = [
+        jq
+        moreutils
+        nodePackages.pnpm
+      ];
 
       installPhase = ''
-        runHook preInstall
-
-        mkdir -p $out
-        mv dist $out
-
-        runHook postInstall
+        export HOME=$(mktemp -d)
+        pnpm config set store-dir $out
+        pnpm install --ignore-scripts
+
+        # Remove timestamp and sort the json files
+        rm -rf $out/v3/tmp
+        for f in $(find $out -name "*.json"); do
+          sed -i -E -e 's/"checkedAt":[0-9]+,//g' $f
+          jq --sort-keys . $f | sponge $f
+        done
       '';
+
+      dontFixup = true;
+      outputHashMode = "recursive";
+      outputHash = "sha256-iQsCgx1J+Rd706jtG7X0/bXPP4GuugCV1T8cUl+xabM=";
     };
```

Comment:
Is `pnpm` reproducible like this? I imagine it's stable for a given version, but I would be interested to know if this output is stable across pnpm versions

If I find time I'll do some testing myself, but I'm admittedly not as familiar with pnpm and I don't know if they guarantee a particular store layout that's byte-for-byte reproducible (after removing timestamps and sorting keys that is, as this implementation seems to do)

Reply:
I don't know if it's stable across pnpm versions. The same trick has been used in [pot](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/misc/pot/default.nix) and looks like it works well. Since the store is versioned, I would expect that the format won't be changed if the store version is not changed. I'm not familiar with pnpm and this trick is borrowed from pot. I thought if the store format is changed in the future we need to pin the pnpm version, as what we do for other packages.

Reply:
Hmmm, I'll try to look at the pnpm documentation and do testing to confirm whether that is the case in the next week or so

Either way though, until we have an official fetcher though it might be best to avoid these custom fetchers if they aren't vetted

Reply:
@lilyinstarlight Kindly ping :). Instead of waiting for a "proper" solution forever, I thought it's better to merge this to unblock the update of this security-essential program.

This workground works well for a long time and I bet it will work much better than a home-made, broken by design "proper" fetcher. And I don't understand why a home-made fetcher is better. If pnpm decides to change the format of the local store, then no matter how we fetch the data, we need to change our code. Instead, we can pin the pnpm version so that the format won't be changed.

Reply:
As for unblocking this PR, would you be open to just switching it back to the package-lock.json approach for now and work on the pnpm fetcher in a new PR?

I do appreciate your work on this!

Reply:
> We've had issues before where even things as dumb as structured parsers (e.g. jq or yq) subtly changed their output and broke any FOD fetchers that were transforming data with them. So when we create fetchers that always output a stable format, we don't have to care when pnpm or jq or anything updates

Good point! If you are worrying about jq changing the output format we can minimize the json after sorting the items.

> Pinning versions sorta helps, but then we're having to maintain old versions for not much reason when we can have purpose-built code where versions no longer matter (edit: to be clear, we do sometimes use original tooling when creating FODs, like with cargo, but those projects also generally explicitly promise a future-proof reproducible output, which I can't tell is the case with pnpm)

It's not easier to maintain a custom fetcher though. It's not enough to output a stable FOD, it must also work. But a home-made fetcher usually doesn't work in some cases. It's hard to write a correct fetcher.

> As for unblocking this PR, would you be open to just switching it back to the package-lock.json approach for now and work on the pnpm fetcher in a new PR?

It should work, yes. But It's generally against the principle that we should avoid adding large lock files into this repo. And it means I need to use different packages from upstream.

Reply:
> Good point! If you are worrying about jq changing the output format we can minimize the json after sorting the items.

Jq's compact output should be fine there, I don't expect that to change

> It's not easier to maintain a custom fetcher though. It's not enough to output a stable FOD, it must also work. But a home-made fetcher usually doesn't work in some cases. It's hard to write a fetcher.

If they have trouble working with the upstream tooling (e.g. in some of the npm problems we've had), it's almost universally because the upstream tooling couldn't work reproducibly in that case anyway

I've unfortunately been personally bitten enough times by custom and complex FODs in derivations like this one that I've become wary :/

> It should work, yes. But It's generally against the principle that we should avoid adding large lock files into this repo. And it means I need to use different packages from upstream.

If you're going to be this against changing it, I can still review this as-is and don't mind much. It's just going to take a bit longer to do so

I'll try to carve out extra time in the next few days to verify FOD reproducibility here (and for pot, which seemed to have been merged without verification...), since unlike input-addressed derivations, FODs have a hard requirement on byte-for-byte reproducibility